### DISCUSSION TITLE
How can I use a Micronaut bean defined with @RequestScope?

### DISCUSSION THREAD

**Discussion by nedelva** [micronaut-projects/micronaut-core#10502] [Q&A]
Say you have something like
```java
@RequestScope
public class MyRequestScopedBean {
    private String message;
    private Locale locale;

    @Inject
    public MyRequestScopedBean() {
        this.message = "This is a request scoped bean";
    }

    public String getMessage() {
        // Use locale if needed
        return message;
    }

    @Inject
    public void setLocale(Locale locale) {
        this.locale = locale;
    }
}
```
The idea is to have this injected in a REST controller method. This way it can benefit from Micronaut's automatic binding of Locale from the HTTP Accept-Language header. But will this work?
```java
@Controller
public class MyController {

   @Get("/message")
   public  String getMessage(Locale locale, MyRequestScopedBean myBean) {
        //a Locale-dependent message
        return myBean.getMessage();
   }
}
```
My hope is that in this way MyRequestScopedBean will be instantiated on every http request, AND also will have its Locale dependency injected.
 
Regarding the testing. A test such as this one:
```java
@MicronautTest
class MyRequestScopedBeanTest {
    @Inject
    MyRequestScopedBean bean;
}
```
won't work, because Micronaut complains (rightfully) that there is no request available.

So, how can I inject the bean in the test class and in the controller?

**Comment by graemerocher** [comment]
inject it in the constructor of the controller

**Reply by nedelva** [reply]
Sure that would be my first choice. But then how is the Locale object - specific to the HTTP request- being injected into the bean?
I imagine it would go like this:

1. A client proxy for MyRequestScopedBean is set at the injection site in the controller;
2. When a http request is handled, the request scope is activated. A new contextual MyRequestScopedBean instance is created and it's dependencies injected (most notably, the Locale object bound at the request scope by the Micronaut)

Is this the correct interpretation? Thing is, for this to work as I want it, the Locale object would need to be put on the request context _before_ the MyRequestScopedBean instance so that it can be injected.

@graemerocher and you said nothing about how to test.

**Reply by sdelamo** [reply]
We have any api to resolve a locale. `HttpLocaleResolver`. 

```java
 index(HttpRequest<?> request) {
        Locale locale = httpLocaleResolver.resolveOrDefault(request);
```
